diff --git a/src/cpio/client_providers/private_key_fetcher_provider/azure/azure_private_key_fetcher_provider.cc b/src/cpio/client_providers/private_key_fetcher_provider/azure/azure_private_key_fetcher_provider.cc
index 0000000..0000001 100644
--- a/src/cpio/client_providers/private_key_fetcher_provider/azure/azure_private_key_fetcher_provider.cc
+++ b/src/cpio/client_providers/private_key_fetcher_provider/azure/azure_private_key_fetcher_provider.cc
@@ -255,11 +255,13 @@
 }
 
 #ifndef TEST_CPIO
-std::unique_ptr<PrivateKeyFetcherProviderInterface>
+absl::Nonnull<std::unique_ptr<PrivateKeyFetcherProviderInterface>>
 PrivateKeyFetcherProviderFactory::Create(
-    HttpClientInterface* http_client,
-    RoleCredentialsProviderInterface* role_credentials_provider,
-    AuthTokenProviderInterface* auth_token_provider) {
+    absl::Nonnull<core::HttpClientInterface*> http_client,
+    absl::Nonnull<RoleCredentialsProviderInterface*>
+        role_credentials_provider,
+    absl::Nonnull<AuthTokenProviderInterface*> auth_token_provider,
+    privacy_sandbox::server_common::log::PSLogContext& log_context) {
   return std::make_unique<AzurePrivateKeyFetcherProvider>(http_client,
                                                           auth_token_provider);
 }

